### PR TITLE
[State Extraction] no migration optimization

### PR DESCRIPTION
**Pull request by zhangchiqing**
This PR made an optimization in state extraction that it will not rebuild the trie if there is no migration. This could save about 22 mins during spork.

### CODE REVIEW

File: ledger/complete/ledger.go [hunk 1]
```diff
@@ -363,61 +363,66 @@ func (l *Ledger) ExportCheckpointAt(
...
+	if noMigration {
+		// when there is no migration, reuse the trie without rebuilding it
+		newTrie = t
+		// when there is no migration, we don't generate the payloads here until later running the
+		// postCheckpointReporters, because the ExportReporter is currently the only
+		// preCheckpointReporters, which doesn't use the payloads.
+	} else {
+		// get all payloads
+		payloads = t.AllPayloads()
+		payloadSize := len(payloads)
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:380]: The diff here is a bit confusing, but there is actually no code change from this line (L380) to L425

File: ledger/complete/ledger.go [hunk 2]
```diff
@@ -363,61 +363,66 @@ func (l *Ledger) ExportCheckpointAt(
...
-		if err != nil {
-			return ledger.State(hash.DummyHash), fmt.Errorf("error applying migration (%d): %w", i, err)
+	var payloads []ledger.Payload
+	var newTrie *trie.MTrie
+
+	noMigration := len(migrations) == 0
+
+	if noMigration {
+		// when there is no migration, reuse the trie without rebuilding it
+		newTrie = t
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:373]: this is the main optimization

File: ledger/complete/ledger.go [hunk 3]
```diff
@@ -467,6 +472,12 @@ func (l *Ledger) ExportCheckpointAt(
 
 	l.logger.Info().Msgf("start running post-checkpoint reporters")
 
+	if noMigration {
+		// when there is no mgiration, we generate the payloads now before
+		// running the postCheckpointReporters
+		payloads = newTrie.AllPayloads()
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:478]: if there is no migration, we delay the reading of the payloads, which is time-consuming, until the checkpoint has been generated

**Review by ramtinms — APPROVED** [review]

**Review by fxamacker — APPROVED** [review]

**Comment by zhangchiqing** [comment]
bors merge

**Pull request merged**